File: mobile/ui/src/commonMain/kotlin/jp/nonbili/meron/ui/QuoteFold.kt
package jp.nonbili.meron.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// Folding a message's quoted tail — the conversation a reply pastes under its
// new text — behind a "•••" toggle, the way Gmail does (and the desktop app,
// see desktop/frontend/src/components/chat/quoteFold.ts). The core finds the
// quote: HTML bodies arrive with it marked by `data-meron-quote`, plain bodies
// with the UTF-16 offset it starts at (`body_quote_start`), which is also how
// Kotlin indexes strings.

/** A plain body split at the core's quote offset; [quote] is empty when there
 *  is nothing to fold. */
internal data class QuotedBody(
    val reply: String,
    val quote: String,
)

internal fun splitQuotedBody(
    body: String,
    quoteStart: Int?,
): QuotedBody =
    if (quoteStart == null || quoteStart <= 0 || quoteStart >= body.length) {
        QuotedBody(body, "")
    } else {
        QuotedBody(body.substring(0, quoteStart).trimEnd(), body.substring(quoteStart).trimEnd())
    }

/** Which messages' quotes the reader opened this session, so a bubble that
 *  scrolls out of the list and back — or whose web view is rebuilt — comes back
 *  the way it was left. Touched from the main thread only. */
internal object QuoteFoldMemory {
    private val open = mutableSetOf<String>()

    fun isOpen(key: String): Boolean = key in open

    fun setOpen(
        key: String,
        value: Boolean,
    ) {
        if (value) open.add(key) else open.remove(key)
    }
}

/** Whether an in-thread search should open the quote: it matches inside it,
 *  by the same case-insensitive rule the highlighter marks with. */
internal fun quoteMatchesSearch(
    quote: String,
    searchQuery: String,
): Boolean {
    val query = searchQuery.trim()
    return quote.isNotEmpty() && query.isNotEmpty() && quote.lowercase().contains(query.lowercase())
}

/** The attribute the core marks an HTML body's quoted tail with. */
internal const val HTML_QUOTE_ATTR = "data-meron-quote"

/** A JavaScript string literal for [value], safe to splice into an inline
 *  script: `<` is escaped so the text can never close the script element. */
internal fun jsStringLiteral(value: String): String =
    buildString {
        append('"')
        value.forEach { char ->
            when (char) {
                '\\' -> append("\\\\")
                '"' -> append("\\\"")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                '<' -> append("\\u003c")
                '\u2028' -> append("\\u2028")
                '\u2029' -> append("\\u2029")
                else -> append(char)
            }
        }
        append('"')
    }

/** The "•••" chip between a plain body's reply and its quoted tail. */
@Composable
internal fun QuoteToggle(
    open: Boolean,
    color: Color,
    onToggle: () -> Unit,
) {
    val label = if (open) tr("chat.hideQuotedText") else tr("chat.showQuotedText")
    val shape = RoundedCornerShape(50)
    Box(
        Modifier
            .padding(vertical = 6.dp)
            .clip(shape)
            .border(1.dp, color.copy(alpha = 0.25f), shape)
            .background(color.copy(alpha = 0.06f))
            .clickable(onClickLabel = label, onClick = onToggle)
            .semantics { contentDescription = label }
            .padding(horizontal = 10.dp, vertical = 2.dp),
    ) {
        Text(
            text = "•••",
            color = color.copy(alpha = 0.65f),
            fontSize = 11.sp,
            lineHeight = 12.sp,
            fontWeight = FontWeight.Bold,
            letterSpacing = 1.sp,
        )
    }
}

File: desktop/frontend/src/components/chat/quoteFold.ts
// Folding a message's quoted tail — the conversation a reply pastes under its
// new text — behind a "•••" toggle, the way Gmail does. The core finds the
// quote: HTML bodies arrive with it marked by `data-meron-quote`, plain bodies
// with the UTF-16 offset it starts at (`body_quote_start`). Which bodies the
// user unfolded is kept for the session, so a bubble that scrolls out of view
// and back comes back the way it was left.

export const HTML_QUOTE_SELECTOR = '[data-meron-quote]'
export const QUOTE_FOLDED_CLASS = 'meron-quote-folded'
export const QUOTE_TOGGLE_CLASS = 'meron-quote-toggle'

const unfolded = new Set<string>()

export function isQuoteUnfolded(key: string): boolean {
  return unfolded.has(key)
}

export function setQuoteUnfolded(key: string, open: boolean) {
  if (open) unfolded.add(key)
  else unfolded.delete(key)
}

/** Split a plain body at the core's quote offset; `quote` is empty when there
 *  is nothing to fold. */
export function splitQuotedBody(body: string, quoteStart: number | null | undefined): { reply: string; quote: string } {
  if (!quoteStart || quoteStart <= 0 || quoteStart >= body.length) return { reply: body, quote: '' }
  return { reply: body.slice(0, quoteStart), quote: body.slice(quoteStart) }
}

export type QuoteToggleLabels = { show: string; hide: string }

/**
 * Fold the marked quote of a frame document behind a toggle placed where the
 * quote starts. Returns the function that folds or unfolds it, or null when the
 * document has no quote. The toggle's dots are drawn by the frame stylesheet
 * rather than written as text, so the in-thread search never matches them.
 */
export function installFrameQuoteFold(
  doc: Document,
  key: string,
  labels: QuoteToggleLabels,
): ((open: boolean) => void) | null {
  const first = doc.querySelector(HTML_QUOTE_SELECTOR)
  if (!first?.parentNode) return null

  // A document wired again gets a fresh toggle rather than a second listener.
  doc.querySelector(`button.${QUOTE_TOGGLE_CLASS}`)?.remove()
  const toggle = doc.createElement('button')
  toggle.type = 'button'
  toggle.className = QUOTE_TOGGLE_CLASS
  first.parentNode.insertBefore(toggle, first)

  const apply = (open: boolean) => {
    setQuoteUnfolded(key, open)
    doc.documentElement.classList.toggle(QUOTE_FOLDED_CLASS, !open)
    const label = open ? labels.hide : labels.show
    toggle.title = label
    toggle.setAttribute('aria-label', label)
    toggle.setAttribute('aria-expanded', String(open))
  }
  toggle.addEventListener('click', (event) => {
    event.preventDefault()
    event.stopPropagation()
    apply(!isQuoteUnfolded(key))
  })
  apply(isQuoteUnfolded(key))
  return apply
}

/** Whether `element` is hidden inside its frame's folded quote. */
export function isInFoldedQuote(element: Element): boolean {
  return (
    element.ownerDocument.documentElement.classList.contains(QUOTE_FOLDED_CLASS) &&
    !!element.closest(HTML_QUOTE_SELECTOR)
  )
}
